keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,423 @@
+
+Filtering is an important part of diagnostics collection. It enables diagnostics
+to be portably sent to third parties that should not have sensitive information
+that logs and files can contain.
+
+By default, KUDO filters all resources (and custom resources) of values
+contained within the KUDO Instance's secrets. This is configurable with the
+`diagnostics.filterSecrets` key.
+
+There may be other fields that need to be filtered. To solve for this, KUDO
+introduces the `diagnostics.bundles.filters` key in `operator.yaml`, which
+contains a list of filters that files pass through before writing to disk.
+Custom filters use either a regular expression or an object reference and
+JSONPath to derive values to filter.
+
+All filtered values appear as `**FILTERED**` in relevant logs and files.
+
+### More Notes
+
+- Do we need to introduce a notion of the collector or controller manager
+  signing and/or encrypting bundles? TBD.
+
+#### Preflight Checks
+
+## Resources
+
+### bundle.resources
+
+An individual bundle resource is represented as a list inside of the
+`diagnostics.bundle.resources` key. Resources ALWAYS have the following keys:
+
+- **name**: The human-readable name of the file.
+- **key**: The machine-readable name of the file. This is used for both
+  references (if needed in the future) and filenames. Extension is OPTIONAL,
+  but may be useful for inferring mime types.
+- **kind**: The kind of bundle item.
+- **spec**: The attributes of a particular kind. This is different for every
+  kind.
+
+Also, may specs may include an `objectRef`. It ALWAYS has the following keys:
+
+- **kind**: The Kubernetes Kind referenced. For example, this may be a
+  Deployment, Pod, StatefulSet, or other resource.
+- **name**: The name of the object. This is a templated field, and has the same
+  template environment as operator templates.
+
+### bundle.resources.Copy
+
+- **path**: Absolute path inside of the referenced pods.
+- **objectRef**
+
+### bundle.resources.Command
+
+- **command**: Command to run. May be a string or an array.
+- **objectRef**
+
+### bundle.resources.Task
+
+- **taskRef**: Name of the task to run. **NOTE**: We MAY need a Pause and Resume
+  task to be able to copy files and run commands during the running of a task.
+  Otherwise, we may want to make this an arbitrary job.
+
+### bundle.resources.HTTP
+
+- **serviceRef**: Object containing references to a Kubernetes service. This is
+  scoped to KUDO-only services.
+- **serviceRef.name**: Name of the service.
+- **serviceRef.port**: Name of the service port. MUST be a named port, not an
+  integer value.
+
+### bundle.filters
+
+Filters are a list of filters. They contain the following keys:
+
+- **name**: The human readable name of the filter.
+- **regex** (optional): Regular expression, not encased in slashes, to use.
+  Regex flags are not supported, and are global and case sensitive.
+- **objectRef** (optional): Required if `jsonPath` is present.
```

Comment:
In this context it would probably make sense to be able to reference CRs, i assume? So we probably need to think about what an objectRef exactly should be

Reply:
Agreed. objectRef, fwiw, is a standard convention. Here we have a bit of a partial one since you're not specifying the full GVK, so I guess we're not technically following convention here 🤔 but we will need to for CRs.